docs/For Developers documentation (#802)
thegetty/quire-docs: content/docs-v1/for-developers.md

  - /documentation/for-developers/
---

Since its launch in December 2022, Quire has been freely available under the BSD 3-Clause License. Quire is a tool for generating full static sites using 11ty, with pages that load quickly in the browser. The Quire CLI provides a straightforward set of commands for building and publishing your site.

## Requirements

> [Node.js](https://nodejs.org/en/) v18 or higher (LTS preferred)

Reviewer: This should be v22 -- see the `engines` property of the CLI and 11ty package file, for now without mentioning LTS. Node's LTS version is 24 currently but I don't know if quire has been sufficiently tested against v24 to document it that way.

  - /documentation/for-developers/
---

Since its launch in December 2022, Quire has been freely available under the BSD 3-Clause License. Quire is a tool for generating full static sites using 11ty, with pages that load quickly in the browser. The Quire CLI provides a straightforward set of commands for building and publishing your site.

## Requirements
[Node.js](https://nodejs.org/en/) v22 preferred

## Repositories

### quire

https://github.com/thegetty/quire

Quire’s primary repository, with the command-line interface and 11ty template packages. 

### quire-docs

https://github.com/thegetty/quire-docs

This repository is specifically for the Quire website and documentation.

## Directory Structure under 11ty
```
  project-root/
  ├── _epub/ # EPUB e-book version of your project
  ├── _includes/ # Language agnostic templates consumed inline by other templates (including layouts) 
  ├── _layouts/ # Chainable language agnostic templates that wrap content via the layout front matter key
  ├── _lib/ # Shared JS utility modules

>   ├── _patches/ # Patch files applied to node_modules automatically via the npm

Reviewer: Could you remove `_patches` here? It's an artifact that will be removed sometime soon (https://github.com/thegetty/quire/issues/1283).

{% q-class "box warning" %}
- Editing `.eleventy.js` can introduce breaking changes. Ensure you read the official 11ty documentation before making any changes.
{% endq-class %}

| Property | Set Via | Expected Value | Description |
| --- | --- | --- | --- |
| `input` | `eleventyConfig.setInputDirectory()` | string | Directory Eleventy scans to resolve templates. Defaults to `content`, or override with the `ELEVENTY_INPUT` environment variable. |
| `output` | `eleventyConfig.setOutputDirectory()` | string | Directory the built site is written to. Defaults to `_site`, or override with `ELEVENTY_OUTPUT`. |
| `data` | `eleventyConfig.setDataDirectory()` | string | Directory (relative to `input`) Eleventy resolves for data files. Defaults to `_computed`, or override with `ELEVENTY_DATA`. |
| `includes` | `eleventyConfig.setIncludesDirectory()` | string | Directory (relative to `input`) for templates consumed inline by other templates. Defaults to `../_includes`, or override with `ELEVENTY_INCLUDES`. |

> | `layouts` | `eleventyConfig.setLayoutsDirectory()` | string | Directory (relative to `input`) Eleventy resolves for layout templates. Defaults to `../_layouts`, or override with `ELEVENTY_LAYOUTS`. |

Reviewer: ```suggestion
| `input` | `ELEVENTY_INPUT` environment variable | string | Directory Eleventy scans to resolve templates. Defaults to `content`. |
| `output` | `ELEVENTY_OUTPUT` environment variable | string | Directory the built site is written to. Defaults to `_site`. |
| `data` | `ELEVENTY_DATA` environment variable | string | Directory (relative to `input`) Eleventy resolves for data files. Defaults to `_computed`. |
| `includes` | `ELEVENTY_INCLUDES` environment variable | string | Directory (relative to `input`) for templates consumed inline by other templates. Defaults to `../_includes`. |
| `layouts` | `ELEVENTY_LAYOUTS` environment variable | string | Directory (relative to `input`) Eleventy resolves for layout templates. Defaults to `../_layouts`. |
```

| `contributor_byline` | boolean | |
| `abstract` | string | Markdown okay. |
| `slug` | url path | Will change the URL of the page. Or use a period `.` to make the URL be the directory name (homepage). Read more in the [*Page Types & Structure*](/docs-v1/pages/#creating-section-landing-pages) chapter of this guide. |
| `toc` | boolean | Default is "true". Page will not display in contents page if "false". |
| `menu` | boolean | Default is "true". Page will not display in menu if "false". |
| `online` | boolean | Default is "true". Page will not display in the online edition if "false". |
| `pdf` | boolean | Default is "true". Page will not display in the PDF edition if "false". |
| `epub` | boolean | Default is "true". Page will not display in either the EPUB or MOBI e-book editions if "false". |
| `image` | url | |

## Shortcodes API
Shortcodes are reusable components that may accept a number of parameters. These shortcodes are used to insert elements into your templates. Refer to the official Eleventy documentation on [shortcodes](https://www.11ty.dev/docs/shortcodes/).

*Supported Template Formats*
- `11ty.js`
- `html`
- `liquid`
- `md`
- `njk`

### Eleventy Config

> Shortcodes may be added via `eleventyConfig.addShortcode` and `eleventyConfig.addPairedShortcode` under the `.eleventy.js` config file. It is highly recommended to add shortcodes via Quire's Shortcode Plugin.

Reviewer: ```suggestion
Shortcodes may be added via `eleventyConfig.addShortcode` and `eleventyConfig.addPairedShortcode` under the `.eleventy.js` config file.
```

*Basic shortcode example*

Javascript
eleventyConfig.addShortcode("year", () => ${new Date().getFullYear()})

> njk

Reviewer: ```suggestion
liquid
```

{% year %}

*Basic paired shortcode example*

Javascript
eleventyConfig.addPairedShortcode("callout", (content, type) => {
return <div class="callout callout--${type}">${content}</div>
})

njk
{% callout "warning" %}
This is important.
{% endcallout %}

### Quire's Shortcode Plugin
Rather than calling eleventyConfig directly, Quire provides a thin wrapper under the `_plugins/shortcodes/index.js` directory. This wrapper (known as `shortcodefactory`) provides page and collection data to your component.

*Why use Quire's Shortcode Plugin*:
- Allows shortcodes to be registered identically across all supported template languages.

> - Decouples shortcode logic from Eleventy's internal structure.

Reviewer: ```suggestion
```

### Creating a shortcode
You should create your shortcodes under the `_plugins/shortcodes` directory.

> After creating your component, register it by calling `addShortcode` or `addPairedShortcode` inside `_plugins/shortcodes/index.js`.

Reviewer: ```suggestion
```

## Configuration

> Quire uses 11ty to generate static site files for publication. Through 11ty's Configuration API, general settings are available for developers in `config.yaml` to configure their Quire project. Learn more about 11ty's Configuration API through their official [documentation](https://www.11ty.dev/docs/config/).

Reviewer: It would be good to distinguish here between 11ty's Configuration API, which is largely executed / configured in  `.eleventy.js` and the `config.yaml` file.

Could you distinguish these two configuration workflows a bit better here by renaming this section "Configuration YAML", and then doing a similar tabular rundown of quire plugins instantiated by `.eleventy.js` undera  section heading "Configuration API"?

| `entryPage` | object | Settings for object/catalogue entry pages. |
| `epub` | object | EPUB output settings. |
| `figures` | object | Figure and image display settings. |
| `footnotes` | object | Footnote display settings. |
| `globalControls` | object | Labels for expand/collapse all controls. |
| `licenseIcons` | boolean | Whether to display license icons. |
| `localization` | object | Localization settings. |
| `menuType` | `full` \| `brief` | Set the appearance of the contents list in the sidebar menu. Setting to `brief` will hide any pages with sub-sections. |
| `navigation` | object | Previous/next navigation button labels. |
| `pageTitle` | object | Page title formatting. |
| `pdf` | object | PDF output settings. |
| `ref` | object | Cross-reference settings. |
| `searchEnabled` | boolean | Enables or disables site search. |
| `tableOfContents` | object | Table of contents display settings. |

## Migrating from Hugo to 11ty

Hugo supports various config file types including YAML, TOML and JSON while 11ty supports YAML with a separate `.eleventy.js` file for Configuration API. 

{% q-class "box warning" %}

> - Editing `.eleventy.js` can introduce breaking changes. Ensure you read the official 11ty documentation before making any changes.

Reviewer: Super important! This belongs up in the "Configuration API" section, however.

{% endq-class %}

### Properties removed from `config.yaml` under Hugo
- `blackfriday`
- `markup`
- `theme`
- `metaDataFormat`
- `pluralizeListTitles`
- `disableKinds`
- `buildDrafts`

> ### Directory Structure under 11ty

Reviewer: This is a great start, it should be toward the top of the document and at a heading level 2. Could you elaborate more on directories used, working from the directories configured in .eleventy.js (eg, public, _site, _includes, etc)?
